### ISSUE TITLE
Writer for .mgf files

### ISSUE THREAD

**Issue by michaelwitting** [rformassspectrometry/MsBackendMgf#4]
What about a write for .mgf files? I guess some code can be refactored from `MSnbase`.

**Comment by lgatto** [comment]
Yes, this will be added - the code in `MSnbase` can be taken (more or less) as is. 

Would we want to be exhaustive in write all the spectra variables we have or only those that are generally present in mgf files? I would vote for the latter and delegate more specialised types of mgf, that expect special header fields, to the specialised backends.

**Comment by michaelwitting** [comment]
The use case I'm having right now would be to have all variable there. Maybe define a similar data frame with key/value-pairs like for the reading. Default is only the core variables and user can change according to the special needs?

**Comment by lgatto** [comment]
And when you mean all variables, is this in the frame of a specific mgf dialect - for example MassBank? Because if so, my suggestion would be for the default `MsBackendMgf` to do the general thing and write only commonly used fields, and for specialised backends to write what is relevant to them.

**Comment by michaelwitting** [comment]
In this case (I'm currently working with) I have MS2 spectra and add some information to them, e.g. IDs of the corresponding MS1 peak, sum formula etc. They should go into the .mgf file, because some other downstream tools (e.g. GNPS ) need to pick them up. But for me it also fine to have a basic version and I can add a specific version of the writer to the code where I need it. 
For MassBank I will also add a writer, once I finished the reading. But MassBank is well defined compared to .mgf.

**Comment by lgatto** [comment]
There is now an `export()` method.

**Comment by jorainer** [comment]
Re `export` - we will have to wait for the Bioconductor submission to go through before merging that into the main `Spectra` branch. Would be good it you could then test the `MsBackendMgf` export functionality to see if that works for you @michaelwitting

**Comment by michaelwitting** [comment]
I will test it soonish...

**Comment by stanstrup** [comment]
I tried to use the data in the example and do an export but I get an empty file:
```r
Spectra::export(be, MsBackendMgf(), file = "temp.mgf")
```

Output:
```
TITLE=msLevel ; retentionTime ; scanNum 
msLevel=
RTINSECONDS=
SCANS=
scanIndex=
mz=
intensity=
centroided=
smoothed=
polarity=
precScanNum=
PEPMASS=
PEPMASSINT=
CHARGE=
collisionEnergy=
isolationWindowLowerMz=
isolationWindowTargetMz=
isolationWindowUpperMz=
```


With my own data I get an error instead;
```r
> Spectra::export(spectras, MsBackendMgf(), file = "temp.mgf")
Fejl i as.vector(x, mode) : 
  coercing an AtomicList object to an atomic vector is supported only for
  objects with top-level elements of length <= 1
```

[spectras.zip](https://github.com/rformassspectrometry/MsBackendMgf/files/5304909/spectras.zip)

**Comment by jorainer** [comment]
This is strange - I will have a look at it. Thanks for reporting @stanstrup

**Comment by jorainer** [comment]
`Spectra::export(be, MsBackendMgf(), file = "temp.mgf")` is not correct, as `be` is not a `Spectra` object but a `MsBackendMgf`. To export the data you would have to first create a `Spectra` object:

```r
sps <- Spectra(be)
export(sps, MsBackendMgf(), file = "temp.mgf")
```

With your own data the story is slightly different. Your `Spectra` has a spectra variable which is a `CharacterList` (column `"isotope"`):

```r
> spectraData(sps)
DataFrame with 217 rows and 21 columns
      msLevel     rtime acquisitionNum scanIndex dataStorage
    <integer> <numeric>      <integer> <integer> <character>
1           1   446.248             NA      1702    <memory>
2           1   410.252             NA      1563    <memory>
...       ...       ...            ...       ...         ...
216         1   507.449             NA      1935    <memory>
217         1   515.389             NA      1966    <memory>
                   dataOrigin centroided  smoothed  polarity precScanNum
                  <character>  <logical> <logical> <integer>   <integer>
1   \\\\science.domain\\gro..       TRUE        NA         1          NA
2   \\\\science.domain\\gro..       TRUE        NA         1          NA
...                       ...        ...       ...       ...         ...
216 \\\\science.domain\\gro..       TRUE        NA         1          NA
217 \\\\science.domain\\gro..       TRUE        NA         1          NA
    precursorMz precursorIntensity precursorCharge collisionEnergy
      <numeric>          <numeric>       <integer>       <numeric>
1            NA                 NA              NA              NA
2            NA                 NA              NA              NA
...         ...                ...             ...             ...
216          NA                 NA              NA              NA
217          NA                 NA              NA              NA
    isolationWindowLowerMz isolationWindowTargetMz isolationWindowUpperMz
                 <numeric>               <numeric>              <numeric>
1                       NA                      NA                     NA
2                       NA                      NA                     NA
...                    ...                     ...                    ...
216                     NA                      NA                     NA
217                     NA                      NA                     NA
       scanNo                 splash     pcgroup                  isotope
    <integer>            <character> <character>          <CharacterList>
1        1636 splash10-0mdi-000019..           1                   ,,,...
2        1502 splash10-03di-090100..          10 ,[11][M]+,[11][M+1]+,...
...       ...                    ...         ...                      ...
216      1860 splash10-0udi-000000..          98                         
217      1890 splash10-000i-090000..          99                         
```
 The `MsBackendMgf` export function can however only handle non-list elements - it's up to the user to eventually *reduce* multiple elements into a single one. Something like.

What you would have to do is to either drop that column before, or convert it to a character:

```r
sps_out <- sps
sps_out$isotope_char <- vapply(sps$isotope, function(z) paste(z, collapse = ";"), character(1))
## Drop that column
sps_out$isotope <- NULL

## Export the data
fl <- tempfile()
export(sps_out, MsBackendMgf(), file = fl)
readLines(fl, n = 20)
 [1] "BEGIN IONS"                                                                                                                                                                                       
 [2] "TITLE=msLevel 1; retentionTime ; scanNum "                                                                                                                                                        
 [3] "msLevel=1"                                                                                                                                                                                        
 [4] "RTINSECONDS=446.2479972"                                                                                                                                                                          
 [5] "scanIndex=1702"                                                                                                                                                                                   
 [6] "centroided=TRUE"                                                                                                                                                                                  
 [7] "polarity=1"                                                                                                                                                                                       
 [8] "scanNo=1636"                                                                                                                                                                                      
 [9] "splash=splash10-0mdi-0000198500-09312cf8d5b2b72892ce"                                                                                                                                             
[10] "pcgroup=1"                                                                                                                                                                                        
[11] "isotope_char=;;;[33][M]+;[33][M+1]+;;[36][M]+;[36][M+1]+;;;[37][M]+;[37][M+1]+;[37][M+2]+;[38][M]+;[38][M+1]+;[38][M+2]+;[20][M]+;[20][M+1]+;[20][M+2]+;;[14][M]+;[14][M+1]+;[15][M]+;[15][M+1]+;"
[12] "438.378662109375 108.779238410169"                                                                                                                                                                
[13] "465.385955810547 75.8970532317633"                                                                                                                                                                
[14] "482.407440185547 314.505523790289"                                                                                                                                                                
[15] "526.431274414062 681.879070357182"                                                                                                                                                                
[16] "527.434265136719 191.699398357088"                                                                                                                                                                
[17] "527.437072753906 134.514823614304"                                                                                                                                                                
[18] "570.458801269531 1355.91087990431"                                                                                                                                                                
[19] "571.460693359375 180.310664817822"                                                                                                                                                                
[20] "571.462463378906 141.124580932719"                                                                                                                                                                
```

**Comment by jorainer** [comment]
In https://github.com/rformassspectrometry/MsBackendMgf/commit/30f879ca59478036d84b4c9be61fc1ecfaea73ae I have added a check that validates the columns of the spectra data, you would then get now a less cryptic error message.

**Comment by stanstrup** [comment]
Thanks! Works.

**Comment by lgatto** [comment]
I believe this issue can be closed since PR #9 has now been merged. Feel free to re-open and follow up otherwise.
